server/service/async/collect.go

```diff
@@ -0,0 +1,151 @@
+package async
+
+import (
+	"context"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	redigo "github.com/gomodule/redigo/redis"
+)
+
+type collector struct {
+	// immutable after creation
+	name         string
+	pool         fleet.RedisPool
+	ds           fleet.Datastore
+	execInterval time.Duration
+	jitterPct    int
+	lockTimeout  time.Duration
+	handler      func(context.Context, fleet.Datastore, fleet.RedisPool, *collectorExecStats) error
+	errHandler   func(string, error)
+
+	// mutable, must be protected by mutex
+	mu    sync.RWMutex
+	stats collectorStats
+}
+
+type collectorStats struct {
+	ExecCount     int
+	FailuresCount int
+
+	MinExecDuration  time.Duration
+	MaxExecDuration  time.Duration
+	LastExecDuration time.Duration
+
+	MinExecKeys  int
+	MaxExecKeys  int
+	LastExecKeys int
+
+	MinExecItems  int
+	MaxExecItems  int
+	LastExecItems int
+}
+
+type collectorExecStats struct {
+	Duration time.Duration
+	Keys     int
+	Items    int
+	Failed   bool
+}
+
+func (c *collector) Start(ctx context.Context) {
+	for {
+		wait := c.nextRunAfter()
+		select {
+		case <-time.After(wait):
+			c.exec(ctx)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *collector) exec(ctx context.Context) {
+	keyLock := fmt.Sprintf(collectorLockKey, c.name)
+	conn := c.pool.ConfigureDoer(c.pool.Get())
+	defer conn.Close()
+
+	if _, err := redigo.String(conn.Do("SET", keyLock, 1, "NX", "EX", int(c.lockTimeout/time.Second))); err != nil {
```

Comment:
`c.lockTimeout/time.Second` you mean `c.lockTimeout.Seconds()`?

Reply:
Ah, yeah well it's similar, the method returns a float, the division an integer without smaller fractions (as a time.Duration). I'll update it to the method call, it's probably clearer in intent.